### ISSUE TITLE
Authentication for client-side applications

### ISSUE THREAD

**Issue by willdurand** [mozilla/addons-server#10491]
Our new project https://github.com/mozilla/addons-code-manager is currently a pure client-side application (no server code at all). This application must talk to the AMO API and users should be mostly authenticated, so most (if not all) API calls have to be authenticated.

In addons-frontend, we use a "frontend auth" cookie that is HTTP only. This cookie is sent by the AMO API to the browser. The browser sends the cookie to the frontend _server_ on every HTTP call. The frontend _server_ is the only piece of frontend JS that is able to read this cookie. The cookie value (i.e. the auth token) is read and then transferred to the _client_ JS code by serializing it (in the redux state). Having the token in our frontend code is important because the API requires an `Authorization` header only. The is no cookie-based authentication mechanism.

That works well in the context of addons-frontend (and likely other 3rd-party clients) but it does not work for addons-code-manager. It is possible to log in and get a "frontend auth" cookie in this project (I have a POC), but there is no way to read this cookie from the addons-code-manager JS code because the cookie is HTTP only. There is also no way to perform authenticated HTTP calls to the API because the API ignores it (during the authentication phase).

**This issue is about discussing and finding a way to securely make authenticated API calls from addons-code-manager to the AMO API.**

I personally believe we could implement a cookie-based authentication mechanism so that we don't need to have access to the cookie/token in addons-code-manager and we don't even need to store the token. This likely requires some extra protections against CSRF, but maybe we could use the [same-site cookie](https://blog.mozilla.org/security/2018/04/24/same-site-cookies-in-firefox-60/) policy now? How does that work with `localhost` for development though?

**Comment by EnTeQuAk** [comment]
So, correct me if I'm wrong here.

The api-token cookie is correctly being set, the client just can't read it because of `HttpOnly`, so far so good.

That cookie would be sent on consequent requests though, so that, in theory, we'll be able to add a 3rd authentication mechanism that simply reads that cookie and uses that to authenticate the user. I wonder though if we could just use the regular django session cookie for that?

**Comment by willdurand** [comment]
> The api-token cookie is correctly being set, the client just can't read it because of `HttpOnly`, so far so good.

Correct. The cookie is read by the server code of addons-frontend: https://github.com/mozilla/addons-frontend/blob/19947d1cf925dd5b80fe13ae1b5c866bc3ec4c7f/src/core/server/base.js#L334-L339

> That cookie would be sent on consequent requests though, so that, in theory, we'll be able to add a 3rd authentication mechanism that simply reads that cookie and uses that to authenticate the user. I wonder though if we could just use the regular django session cookie for that?

I am not sure about the Django question but the rest is correct. CSRF would be an issue though, but it could be fixed by using the same-site cookie policy, assuming there is a way to make it work for local dev.

**Comment by EnTeQuAk** [comment]
Hold on, if the frontend-api token is already sent to the API on every request, we could simply re-use that and can keep the `HttpOnly` flag unless I'm mistaken.

Regarding `Same-Site` cookie policy, I wondered if CORS headers would mitigate local-development problems but it seems that it doesn't from what I'm reading in the specs :disappointed:

**Comment by diox** [comment]
> I am not sure about the Django question but the rest is correct. CSRF would be an issue though, but it could be fixed by using the same-site cookie policy, assuming there is a way to make it work for local dev.

`Same-Site` only protects modern clients though. If the API starts accepting the frontend auth cookie as an authentication method, then it opens us to potential CSRF issues on older clients.

BTW, it's currently annoying to support `Same-Site` cookies with the version of Django we have. See https://github.com/mozilla/addons-server/issues/8117
